<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:include filename="inertial_macros.xacro"/>

    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>

    <material name="orange">
        <color rgba="1 0.3 0.1 1"/>
    </material>

    <material name="blue">
        <color rgba="0.2 0.2 1 1"/>
    </material>

    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>

    <material name="red">
        <color rgba="1 0 0 1"/>
    </material>

    <!-- BASE LINK -->

    <link name="base_link"></link>

    <joint name="base_footprint_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <link name="base_footprint"></link>

    <!-- CHASSIS LINK -->

    <joint name="chassis_joint" type="fixed">
        <parent link="base_link"/>
        <child link="chassis_link"/>
        <origin xyz="0.0508 0 0" rpy="0 0 0"/> <!-- Offset the chassis slightly back by 5.08cm-->
    </joint>

    <link name="chassis_link">
        <visual>
            <origin xyz="-0.13462 0 0.06096" rpy="0 0 0"/>
            <geometry>
                <box size="0.26924 0.1524 0.12192"/> 
            </geometry>
            <material name="white"/>
        </visual>
        <collision>
            <origin xyz="-0.13462 0 0.06096" rpy="0 0 0"/>
            <geometry>
                <box size="0.26924 0.1524 0.12192"/> 
            </geometry>
        </collision>
        <xacro:inertial_box mass="0.5" x="0.26924" y="0.1524" z="0.12192">
            <origin xyz="-0.13462 0 0.06096" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>

    <!-- LEFT WHEEL LINK -->

    <joint name="left_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="left_wheel_link"/>
        <origin xyz="0 -0.09144 -0.0127" rpy="-1.5707963267948966 0 0"/> <!-- Rotate wheel -->
        <axis xyz="0 0 -1"/>
    </joint>

    <link name="left_wheel_link">
        <visual>
            <geometry>
                <cylinder radius="0.0325" length="0.027"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.0325" length="0.027"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.027" radius="0.035">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <!-- RIGHT WHEEL LINK -->

    <joint name="right_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="right_wheel_link"/>
        <origin xyz="0 0.09144 -0.0127" rpy="1.5707963267948966 0 0"/> <!-- Rotate wheel -->
        <axis xyz="0 0 1"/>
    </joint>

    <link name="right_wheel_link">
        <visual>
            <geometry>
                <cylinder radius="0.0325" length="0.027"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.0325" length="0.027"/>
            </geometry>
        </collision>
        <xacro:inertial_cylinder mass="0.1" length="0.027" radius="0.035">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <!-- CASTER WHEEL LINK -->

    <joint name="caster_wheel_joint" type="fixed">
        <parent link="base_link"/>
        <child link="caster_wheel_link"/>
        <origin xyz="-0.15 0 0" rpy="0 0 0"/>
    </joint>

    <link name="caster_wheel_link">
        <visual>
            <geometry>
                <sphere radius="0.04572"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.04572"/>
            </geometry>
            <material name="black"/>
        </collision>
        <xacro:inertial_sphere mass="0.1" radius="0.04572">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_sphere>
    </link>

    <!-- LIDAR LINK -->

    <joint name="lidar_joint" type="fixed">
        <parent link="base_link"/>
        <child link="lidar_link"/>
        <origin xyz="-0.06 0 0.13442" rpy="0 0 0"/>
    </joint>

    <link name="lidar_link">
        <visual>
            <geometry>
                <cylinder radius="0.035" length="0.025"/>
            </geometry>
            <material name="red"/>
        </visual>
        <xacro:inertial_cylinder mass="0.140" length="0.025" radius="0.035">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <ros2_control name="GazeboSimSystem" type="system">
        <hardware>
            <plugin>gz_ros2_control/GazeboSimSystem</plugin>
        </hardware>
        <joint name="left_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-10</param>
                <param name="max">10</param>
            </command_interface>
            <state_interface name="velocity"/>
            <state_interface name="position"/>
        </joint>
        <joint name="right_wheel_joint">
            <command_interface name="velocity">
                <param name="min">-10</param>
                <param name="max">10</param>
            </command_interface>
            <state_interface name="velocity"/>
            <state_interface name="position"/>
        </joint>
    </ros2_control>

    <gazebo>
        <plugin filename="gz_ros2_control-system" name="gz_ros2_control::GazeboSimROS2ControlPlugin">
            <parameters>$(find chassis)/config/my_controllers.yaml</parameters>
        </plugin>
    </gazebo>

    <!-- <gazebo>
        <plugin name="gazebo_ros2_control" filename="libgazebo_ros2_control.so"/>
    </gazebo> -->

</robot>